### ISSUE TITLE
unsupported locale setting error in 2.1.0

### ISSUE THREAD

**Issue by thornycrackers** [urwid/urwid#377]
##### Description:
I'm trying to import pudb into my code base with `from pudb import set_trace; set_trace()` and I get this error. 

```
unsupported locale setting error

/opt/venv/lib/python3.6/site-packages/urwid/util.py in detect_encoding
line 58:         locale.setlocale(locale.LC_ALL, initial)
initial | ('en_US', 'UTF-8')
locale | <module 'locale' from '/usr/local/lib/python3.6/locale.py'>


/usr/local/lib/python3.6/locale.py in setlocale
line 598:     return _setlocale(category, locale)
category | 6
locale | 'en_US.UTF-8'
```

If I downgrade to `2.0.1` everything is fine. This is running inside the `python:3.6-slim-stretch` image.
 

##### Affected versions (if applicable)
2.1.0

**Comment by tonycpsu** [comment]
Likely related to #259 , #279 .

**Comment by hoopes** [comment]
I opened a stack overflow for this (before i saw this ticket...): 

https://stackoverflow.com/questions/59295250/python-default-locale-unsupported-locale-setting/59305150#59305150

The locale error from the SO issue was being thrown from urwid when using pudb.

**Comment by pieqq** [comment]
I'm using urwid for a TUI program I'm writing, and I got the same issue when trying to run my program on Ubuntu Core 16.

Here is the output of `locale -a` and `locale` in UC16:

```
$ locale -a
C
C.UTF-8
POSIX

$ locale
locale: Cannot set LC_ALL to default locale: No such file or directory
LANG=C.UTF-8
LANGUAGE=
LC_CTYPE="C.UTF-8"
LC_NUMERIC=en_US.UTF-8
LC_TIME=en_US.UTF-8
LC_COLLATE="C.UTF-8"
LC_MONETARY=en_US.UTF-8
LC_MESSAGES="C.UTF-8"
LC_PAPER=en_US.UTF-8
LC_NAME=en_US.UTF-8
LC_ADDRESS=en_US.UTF-8
LC_TELEPHONE=en_US.UTF-8
LC_MEASUREMENT=en_US.UTF-8
LC_IDENTIFICATION=en_US.UTF-8
LC_ALL=
```

Because of this, when urwid tries to execute the [code that was added for 2.1.0 in util.py](https://github.com/urwid/urwid/pull/279/files#diff-f2004e2753aca67c165b2323b2ef58d4), it fails. This issue can be replicated manually:

1. Setup a device with Ubuntu Core 16 on it. You can also [create a Linux container running UC16](https://ubuntu.com/blog/ubuntu-core-in-lxd-containers) or something similar.
2. Log into your UC16 device/container, then:

```
$ python3
Python 3.6.9 (default, Nov  7 2019, 10:44:02)
[GCC 8.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import locale
>>> locale.getlocale()
('en_US', 'UTF-8')
>>> initial = locale.getlocale()
>>> locale.setlocale(locale.LC_ALL, initial)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/snap/qabro/102/usr/lib/python3.6/locale.py", line 598, in setlocale
    return _setlocale(category, locale)
locale.Error: unsupported locale setting
```

**Comment by GuillaumeSeren** [comment]
Hello,
I run into this issue while testing a package for gentoo on a chroot environment.

The issue append on the chroot env,
and append with 'urwid-2.1.0' and 'urwid-2.1.1'

During the test of alot package
```
ERROR: tests.widgets.test_globals (unittest.loader._FailedTest)
----------------------------------------------------------------------
ImportError: Failed to import test module: tests.widgets.test_globals
Traceback (most recent call last):
  File "/usr/lib/python3.7/unittest/loader.py", line 436, in _find_test_path
    module = self._get_module_from_name(name)
  File "/usr/lib/python3.7/unittest/loader.py", line 377, in _get_module_from_name
    __import__(name)
  File "/var/tmp/portage/mail-client/alot-0.9.1/work/alot-0.9.1/tests/widgets/test_globals.py", line 22, in <module>
    from alot.widgets import globals as globals_
  File "/var/tmp/portage/mail-client/alot-0.9.1/work/alot-0.9.1/alot/widgets/globals.py", line 10, in <module>
    import urwid
  File "/usr/lib/python3.7/site-packages/urwid/__init__.py", line 26, in <module>
    from urwid.widget import (FLOW, BOX, FIXED, LEFT, RIGHT, CENTER, TOP, MIDDLE,
  File "/usr/lib/python3.7/site-packages/urwid/widget.py", line 27, in <module>
    from urwid.util import (MetaSuper, decompose_tagmarkup, calc_width,
  File "/usr/lib/python3.7/site-packages/urwid/util.py", line 61, in <module>
    detected_encoding = detect_encoding()
  File "/usr/lib/python3.7/site-packages/urwid/util.py", line 58, in detect_encoding
    locale.setlocale(locale.LC_ALL, initial)
  File "/usr/lib/python3.7/locale.py", line 608, in setlocale
    return _setlocale(category, locale)
locale.Error: unsupported locale setting
```

The issue is fixed by that patch fd010f8336407ad1a42522b424d2ea7b7800ebc4
by jspricke https://github.com/jspricke/urwid/commit/fd010f8336407ad1a42522b424d2ea7b7800ebc4
```diff
--- a/urwid/util.py
+++ b/urwid/util.py
@@ -55,7 +55,10 @@ def detect_encoding():
         else:
             raise
     finally:
-        locale.setlocale(locale.LC_ALL, initial)
+        try:
+            locale.setlocale(locale.LC_ALL, initial)
+        except locale.Error:
+            pass

 if 'detected_encoding' not in locals():
     detected_encoding = detect_encoding()
```

You would like to see fixed, let me know if I can test / help.
